shunting-yard.h

```diff
@@ -189,17 +192,30 @@ struct rpnBuilder {
 
   // Check if a character is the first character of a variable:
   static inline bool isvarchar(const char c) {
-    return isalpha(c) || c == '_';
+    return  isUTF8char(c) || isalpha(c) || c == '_';
+  }
+
+  static inline bool isUTF8char(const char c) {
+    if(unicodeCharCounter > 0) return --unicodeCharCounter;
+    if(c & 0x80) { // This is the start of a unicode character
+      ++unicodeCharCounter;
+      if(c & 0x40) ++unicodeCharCounter;
+      else return true;
+      if(c & 0x20) ++unicodeCharCounter;
+      else return true;
+      if(c & 0x10) ++unicodeCharCounter;
+      else return true;
+    }
+    return false;
   }
```

Comment:
There is one important problem with this implementation, a function called `is_anything` semantically should only return a boolean value without making any collateral state changes. In this case, I am more worried about the `isvarchar()` function that is already used in other places and would start to cause unexpected collateral effects, and we would report no errors because we would not detect anything wrong.

I believe we should be able to reason about the start and end of a Unicode character in the context where we are actually parsing a variable name, i.e. inside the `parseVar()` func below, while the `isvarchar` function should return true only if it is a letter, `_` or if it is the very first byte of an Unicode character.

Reply:
One option would be to split this "isUTF8char" in 2: One that checks if it is the first char of a Unicode character, and another that would parse an entire UTF8 character and return the ending position using a pointer to `char*`, e.g.:

```
static inline bool parseUTF8char(const char* c, const char** rest) {
    // ...
}
```

And then use this function inside `parseVar()` if we detect that the first character is the first character of an UTF8 character. We should be able to detect invalid encodings inside parseUTF8char and throw an exception if this is the case.

Reply:
Good idea. I also didn't like the global variable implementation and it is a rather naïve implementation as well. I'll look into it when I have some time